docs/content/en/docs/Design/do_transformation.md

```diff
@@ -0,0 +1,227 @@
+```mlir
+func.func @my_function(%value: i32, %inputIndex: index) -> i32 {
+  ...
+  // Violation: for-loop uses %inputIndex as upper bound which causes a input-dependent control-flow
+  %result = scf.for %iv = %begin to %inputIndex step %step_value iter_args(%arg1 = %value) -> i32 {
+    %output = arith.muli %arg1, %arg1 : i32
+    scf.yield %output : i32
+  }
+  ...
+}
+
+// After applying Loop-Transformation
+func.func @my_function(%value: i32, %inputIndex: index) -> i32 {
+  ...
+  // %MAX_INDEX is a constant that defines the maximum possible index value
+  %result = scf.for %iv = %begin to %MAX_INDEX step %step_value iter_args(%arg1 = %value) -> i32 {
+    %output = arith.muli %arg1, %agr1 : i32
+    %cond = arith.cmpi eq, %iv, %inputIndex : index
+    %newOutput = arith.select %cond, %output, %arg1
+    scf.yield %newOutput : i32
+    }
+  ...
+}
+```
+- `scf.while`: This operation represents a generic while/do-while loop that keeps iterating as long as a condition is met. An input-dependent while condition introduces a data-dependent control flow that violates data-oblivious constraints. It's challenging to implement early-exits since the scf.while is also expected to [terminate with `scf.yield`](https://mlir.llvm.org/docs/Dialects/SCFDialect/#scfwhile-scfwhileop).
+
+```mlir
+// Before applying Loop-Transformation
+func.func @my_function(%input: i16){
+  %zero = arith.constant 0 : i16
+  %result = scf.while (%arg1 = %input) : (i16) -> i16 {
+    %cond = arith.cmpi slt, %arg1, %zero : i16
+    // Violation: scf.while uses %cond whose value depends on %input
+    scf.condition(%cond) %arg1 : i16
+  } do {
+  ^bb0(%arg2: i16):
+    %mul = arith.muli %arg2, %arg2: i16
+    scf.yield %mul
+  }
+  ...
+  return
+}
+
+// After applying Loop-Transformation
+func.func @my_function(%input: i16){
+  %zero = arith.constant 0 : i16
+  %begin = arith.constant 1 : index
+  ...
+  // Replace while-loop with a for-loop with a constant bound %MAX_ITER
+  %result = scf.for %iv = %begin to %MAX_ITER step %step_value iter_args(%iter_arg = %input) -> i16 {
+    %cond = arith.cmpi slt, %iter_arg, %zero : i16
+    %mul = arith.muli %iter_arg, %iter_arg : i16
+    %output = arith.select %cond, %mul, %iter_arg
+    scf.yield %output
+  }
+  ...
+  return
+}
+
+```
+### (3) Access-Transformation
+Input-dependent memory access cause data-dependent memory footprints. A naive data-oblivious solution to this maybe doing read-write operations over the entire data structure while only performing the desired save/update operation for the index of interest. For simplicity, we only look at load/store operations for tensors as they are well supported structures in high-level MLIR likely emitted by most frontends. We drafted the following non-SIMD and SIMD-friendly approaches for this transformation:
```

Comment:
remove the SIMD-friendly version, as we decided against implementing that (instead hoping that vectorization will be able to handle it)